### PR TITLE
Autocompletion

### LINKED ISSUE #5: Tab completion
Handle a [complete_request](http://ipython.org/ipython-doc/dev/development/messaging.html#completion) to make tab completion work.

### PR DESCRIPTION
**Pull request by flying-sheep**
depends on #63, fixes #5

maybe in the far future we can use [this monster](https://github.com/rstudio/rstudio/blob/master/src/cpp/session/modules/SessionRCompletions.R) instead…

### CODE REVIEW

**Comment by takluyver** [comment]
Awesome, thanks! It looks like this should be easy to adapt to the msgspec5 request - I hope to make that change soon.

File: R/kernel.r [hunk 1]
```diff
@@ -99,11 +99,43 @@ handle_shell = function() {
...
+    code <- request$content$line
+    
+    c.info <- with_top_env(executor$userenv, {
+        utils:::.assignLinebuffer(code)
+        utils:::.assignEnd(request$content$cursor_pos)
+        utils:::.guessTokenFromLine()
+        utils:::.completeToken()
+        
+        c(list(comps = utils:::.retrieveCompletions()),
+          utils:::.guessTokenFromLine(update = FALSE))
```
Comment (takluyver) [inline R/kernel.r]: I was a bit puzzled by what was going on here - is it worth a comment to explain that .guessTokenFromLine with update=FALSE returns the start and token fields? (it also surprised me that `update` changes the return type).

Comment (flying-sheep) [inline R/kernel.r]: ha! yeah sorry: that’s exactly the perfect case for a comment.

yes: all those functions act on an environment called `.CompletionEnv`, where they set a few variables.

maybe we could pull the variables out of the `.CompletionEnv` directly, but i found that calling this function with `update = FALSE` does exactly that. (and i’ve not found how to access `.CompletionEnv`)

**Comment by lucacerone** [comment]
I tried it on my Jupyter install with python 2 and it seems to work like a champ!

**Comment by flying-sheep** [comment]
@takluyver would you pull it?

i addressed your concern about comments

**Comment by takluyver** [comment]
Thanks! Sorry it took me a while to get to this - this week has been the IPython dev meeting, so I've been pretty busy.

**Comment by flying-sheep** [comment]
no worries :)

great this got in!

**Pull request merged**